File: golang/0345.go
func reverseVowels(s string) string {
	length := len(s)
	res := []byte(s)
	n := length - 1

	for i, c := range s {
		if n < i {
			break
		}
		if vowelCheck(uint8(c)) {
			for j := n; j > i; j-- {
				if vowelCheck(s[j]) {
					res[i] = s[j]
					res[j] = s[i]
					n = j - 1
					break
				} else {
					res[j] = s[j]
				}
			}
		} else {
			res[i] = byte(c)
		}
	}
	return string(res)
}

func vowelCheck(c uint8) bool {
	if c == 'a' || c == 'i' || c == 'e' || c == 'u' || c == 'o' ||
		c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U' {
		return true
	}
	return false
}

File: rust/0345.rs
impl Solution {
    pub fn reverse_vowels(s: String) -> String {
        fn is_vowel(c: char) -> bool {
            "aeiouAEIOU".contains(c)
        }

        let mut res: Vec<char> = s.chars().collect();
        let mut left = 0;
        let mut right = res.len() - 1;

        while left < right {
            while left < res.len() && !is_vowel(res[left]) {
                left += 1;
            }
            while right > 0 && !is_vowel(res[right]) {
                right -= 1;
            }
            if left < right {
                res.swap(left, right);
                left += 1;
                right += 1;
            }
        }

        res.into_iter().collect()
    }
}
